tools/ts_library_builder/build_library.ts

```diff
@@ -0,0 +1,370 @@
+  NamespaceDeclarationKind,
+  Project,
+  ScriptTarget,
+  SourceFile,
+  Type,
+  TypeGuards
+} from "ts-simple-ast";
+import {
+  addInterfaceProperty,
+  addSourceComment,
+  addVariableDeclaration,
+  flattenNamespace,
+  getSourceComment,
+  loadDtsFiles,
+  loadFiles,
+  namespaceSourceFile
+} from "./ast_util";
+
+export interface BuildLibraryOptions {
+  /**
+   * The path to the root of the deno repository
+   */
+  basePath: string;
+
+  /**
+   * The path to the current build path
+   */
+  buildPath: string;
+
+  /**
+   * Denotes if the library should be built with debug information (comments
+   * that indicate the source of the types)
+   */
+  debug?: boolean;
+
+  /**
+   * The path to the output library
+   */
+  outFile: string;
+
+  /**
+   * Execute in silent mode or not
+   */
+  silent?: boolean;
+}
+
+/**
+ * A preamble which is appended to the start of the library.
+ */
+// tslint:disable-next-line:max-line-length
+const libPreamble = `// Copyright 2018 the Deno authors. All rights reserved. MIT license.
+
+// Note that \`@internal/*\` modules are internal to Deno and can only be used
+// for their type information.
+
+/// <reference no-default-lib="true" />
+/// <reference lib="esnext" />
+
+`;
+
+// The path to the msg_generated file relative to the build path
+const MSG_GENERATED_PATH = "/gen/msg_generated.ts";
+
+// An array of enums we want to expose pub
+const MSG_GENERATED_ENUMS = ["ErrorKind"];
+
+/**
+ * Generate the runtime library for Deno and write it to the supplied out file
+ * name.
+ */
+export function main({
+  basePath,
+  buildPath,
+  debug,
+  outFile,
+  silent
+}: BuildLibraryOptions) {
+  if (!silent) {
+    console.log("-----");
+    console.log("build_lib");
+    console.log();
+    console.log(`basePath: "${basePath}"`);
+    console.log(`buildPath: "${buildPath}"`);
+    console.log(`debug: ${!!debug}`);
+    console.log(`outFile: "${outFile}"`);
+    console.log();
+  }
+
+  // the inputProject will take in the TypeScript files that are internal
+  // to Deno to be used to generate the library
+  const inputProject = new Project({
+    compilerOptions: {
+      baseUrl: basePath,
+      declaration: true,
+      emitDeclarationOnly: true,
+      module: ModuleKind.AMD,
+      moduleResolution: ModuleResolutionKind.NodeJs,
+      noLib: true,
+      paths: {
+        "*": ["*", `${buildPath}/*`]
+      },
+      preserveConstEnums: true,
+      strict: true,
+      stripInternal: true,
+      target: ScriptTarget.ESNext
+    }
+  });
+
+  // Add the input files we will need to generate the declarations, `globals`
+  // plus any modules that are importable in the runtime need to be added here
+  // plus the `lib.esnext` which is used as the base library
+  inputProject.addExistingSourceFiles([
+    `${basePath}/node_modules/typescript/lib/lib.esnext.d.ts`,
+    `${basePath}/js/deno.ts`,
+    `${basePath}/js/globals.ts`
+  ]);
+
+  // TODO check diagnostics for any issues
```

Comment:
Will diagnostics be printed to stdout? It would be good to have a simple assert here that fails if there are any diagnostics.

Reply:
Note: The semantic, syntactic, global, options, and declaration (since `declaration: true`) diagnostics are available on `emitResult.getDiagnostics()`.

Generally, diagnostics can be retrieved by calling `inputProject.getPreEmitDiagnostics()`, which returns the semantic, syntactic, global, config file parsing, options, and if enabled the declaration diagnostics. For specific diagnostics, they can be retrieved from the program (ex. `inputProject.getProgram().getSemanticDiagnostics()`).